src/client/components/DicePanel.tsx

```diff
@@ -1,57 +1,382 @@
-            onClick={() => setActive(index)}
+            className={"tab-button"}
+            onClick={() => {
+              addTemplateCategory();
+            }}
+            title="Add Category"
           >
-            <FontAwesomeIcon icon={icon} fixedWidth />
+            <FontAwesomeIcon icon={faPlus} fixedWidth />
           </div>
-        ))}
+        )}
       </div>
 
-      <div className={clsx("tab", active === 0 ? "active" : "")}>
-        <DiceInterface />
-        <DiceInput />
-      </div>
-      {diceCategories.map((icon, index) => (
-        <div
-          key={icon.iconName}
-          className={clsx("tab", active === index + 1 ? "active" : "")}
-        >
-          <DiceTemplates categoryIndex={index} />
-        </div>
-      ))}
+      {renderContent(active)}
     </div>
   );
 });
+
+function DiceTemplateButton({
+  category,
+  active,
+  onSetActive,
+}: {
+  category: RRDiceTemplateCategory;
+  active: boolean;
+  onSetActive: () => void;
+}) {
+  const [addMenuVisible, setAddMenuVisible] = useState(false);
+
+  return (
+    <Popover
+      interactive
+      onClickOutside={() => setAddMenuVisible(false)}
+      visible={addMenuVisible}
+      placement="bottom"
+      content={<DiceTemplateCategoryEditor category={category} />}
+    >
+      <div
+        className={clsx("tab-button", active ? "active" : "")}
+        onClick={(e) => {
+          e.stopPropagation();
+          if (e.button === 0) onSetActive();
+          else if (e.button === 2) {
+            e.stopPropagation();
+          }
+        }}
+        onContextMenu={(e) => {
+          e.stopPropagation();
+          e.preventDefault();
+          if (
+            fixedCategoryIcons.findIndex(
+              (fixedIcon) => fixedIcon === category.icon
+            ) > -1
+          )
+            return;
+          setAddMenuVisible(true);
+        }}
+        title={category.categoryName}
+      >
+        <FontAwesomeIcon icon={iconMap[category.icon]} fixedWidth />
+      </div>
+    </Popover>
+  );
+}
+
+function DiceTemplateCategoryEditor({
+  category,
+}: {
+  category: RRDiceTemplateCategory;
+}) {
+  const myself = useMyProps("diceTemplateCategories", "id");
+  const dispatch = useServerDispatch();
+
+  function deleteCategory() {
+    dispatch({
+      actions: [
+        playerDeleteDiceTemplateCategory({
+          id: myself.id,
+          categoryId: category.id,
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
+    });
+  }
+
+  function updateCategory(
+    category: RRDiceTemplateCategory,
+    newIcon?: typeof categoryIcons[number],
+    newName?: string
+  ) {
+    dispatch({
+      actions: [
+        playerUpdateDiceTemplateCategory({
+          id: myself.id,
+          category: {
+            id: category.id,
+            changes: {
+              icon: newIcon ?? category.icon,
```

Comment:
This will unnecessarily overwrite the icon with the existing icon if no new icon is set. This might create unnecessary edge cases. We could update the function signature to something like this instead: `categoryId: RRDiceTemplateCategoryID, changes: Partial<Pick<RRDiceTemplateCategory, "icon" | "categoryName">>` and then simply hand over the changes here.